# 3. SQL-анализ и продуктовые метрики

**Цель:** перенести данные в реляционную среду PostgreSQL для расчёта сложных продуктовых метрик и подготовки данных к визуализации в Superset.

**Этапы:**
- загрузка данных в PostgreSQL;
- расчёт unit-экономики и продуктовых метрик;
- сегментация пользователей;
- когортный анализ, исследование Retention (удержания) пользователей по месяцам их регистрации (первой поездки в сезоне);
- формирование аналитических выводов.

---

### Импорт библиотек и подключение к базе данных

In [1]:
import pandas as pd
import os

from dotenv import load_dotenv
from urllib.parse import quote_plus

# импорт модуля для создания соединения с базой
from sqlalchemy import create_engine

In [2]:
# настройка отображения SQL-результатов в виде таблицы

import prettytable
prettytable.DEFAULT = prettytable.TableStyle

In [3]:
# загрузим расширение ipython-sql для выполнения SQL-запросов прямо в ноутбуке
%load_ext sql

Подключимся к PostgreSQL из ноутбука:

In [4]:
load_dotenv()

user = os.getenv("DB_USER")
password = quote_plus(os.getenv("DB_PASS"))
host = os.getenv("DB_HOST", "localhost")
port = int(os.getenv("DB_PORT", 5432))
database = os.getenv("DB_NAME")

# создадим соединение
engine = create_engine(f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}")
conn_str = f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}"

# проверим соединение
with engine.connect() as conn:
    print("Успешное подключение к PostgreSQL")

Успешное подключение к PostgreSQL


In [5]:
data_path = "../data/clean/kicksharing_clean.parquet"

df = pd.read_parquet(data_path)

df.to_sql("rides", engine, index=False, if_exists="replace")

print("Таблица rides успешно загружена в PostgreSQL")

Таблица rides успешно загружена в PostgreSQL


In [6]:
# подключимся к созданной базе данных через ipython-sql
conn_str = f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}"
%sql $conn_str

Проверим работу SQL-запроса:

In [7]:
%%sql
select * from rides limit 5;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
5 rows affected.


[(252618395, 517007040, 'M', 26, 'GRD', 'UNM', 'Г МОСКВА', 7.49, 50.0, 300.0, 'SL', 0.06978747, datetime.datetime(2024, 4, 17, 10, 43, 29), datetime.datetime(2024, 4, 17, 10, 43, 34, 133897), datetime.datetime(2024, 4, 17, 10, 45, 22, 547000), 3, datetime.datetime(2024, 4, 17, 10, 43, 34, 133897), 64.98, 0.0, 0.0, 1.806885, 2.317385, 0.0, 0.0, 10, 2, 4),
 (252614779, 517007040, 'M', 26, 'GRD', 'UNM', 'Г МОСКВА', 7.99, 50.0, 300.0, 'SL', 0.8056608, datetime.datetime(2024, 4, 17, 16, 45, 16), datetime.datetime(2024, 4, 17, 16, 45, 20, 641980), datetime.datetime(2024, 4, 17, 16, 51, 26, 403392), 3, datetime.datetime(2024, 4, 17, 16, 45, 20, 641980), 97.94, 0.0, 0.0, 6.0960236, 7.9297013, 0.0, 0.0, 16, 2, 4),
 (252632694, 269139778, 'F', 30, 'GRD', 'UNM', 'Г МОСКВА', 7.99, 50.0, 300.0, 'SL', 0.04636136, datetime.datetime(2024, 4, 17, 16, 47, 59), datetime.datetime(2024, 4, 17, 16, 48, 4, 538852), datetime.datetime(2024, 4, 17, 16, 49, 47, 783760), 3, datetime.datetime(2024, 4, 17, 16, 48, 4, 538852), 65.98, 0.0, 0.0, 1.7207484, 1.6165534, 0.0, 0.0, 16, 2, 4),
 (252613575, 269139778, 'F', 30, 'GRD', 'UNM', 'Г МОСКВА', 7.99, 50.0, 300.0, 'SL', 0.06449076, datetime.datetime(2024, 4, 17, 16, 50, 58), datetime.datetime(2024, 4, 17, 16, 51, 1, 965350), datetime.datetime(2024, 4, 17, 16, 52, 32, 398000), 3, datetime.datetime(2024, 4, 17, 16, 51, 1, 965350), 65.98, 0.0, 0.0, 1.5072109, 2.5672886, 0.0, 0.0, 16, 2, 4),
 (252629528, 21817314, 'F', 30, 'GRD', 'UNM', 'МОСКОВСКАЯ ОБЛ', 7.49, 50.0, 300.0, 'SL', 0.055214338, datetime.datetime(2024, 4, 19, 10, 32, 8), datetime.datetime(2024, 4, 19, 10, 32, 15, 964200), datetime.datetime(2024, 4, 19, 10, 38, 11, 978319), 3, datetime.datetime(2024, 4, 19, 10, 32, 15, 964200), 94.94, 0.0, 0.0, 5.9335685, 0.5583251, 0.0, 0.0, 10, 4, 4)]

---

### Расчёт базовых метрик и unit-экономики

#### Количество поездок и пользователей

In [8]:
%%sql
select
    count(distinct order_rk) as total_rides,
    count(distinct party_rk) as total_users
from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(357110, 57532)]

#### Период поездок

In [9]:
%%sql
select
    concat('C ', min(book_start_dttm::date), ' по ', max(book_start_dttm::date)) as season
from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[('C 2024-04-17 по 2024-10-21',)]

#### Статистики по поездкам

In [10]:
%%sql
select
    round(avg(distance_km)::numeric, 3) as avg_distance_km,
    percentile_cont(0.5) within group (order by distance_km) as median_distance_km,
    round(avg(trip_duration_min)::numeric, 2) as avg_trip_duration,
    percentile_cont(0.5) within group (order by trip_duration_min) as median_trip_duration,
    round(avg(avg_speed_kmh)::numeric, 2) as avg_speed_kmh,
    percentile_cont(0.5) within group (order by avg_speed_kmh) as median_speed_kmh
from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('2.428'), 1.769460678100586, Decimal('11.43'), 8.033922672271729, Decimal('13.53'), 13.022547721862793)]

- Средняя дистанция: 2,43 км
- Медианная дистанция: 1,77 км
- Средняя длительность: 11,43 мин
- Медианная длительность: 8,03 мин
- Средняя скорость: 13,53 км/ч
- Медианная скорость: 13,02 км/ч

#### RPU (rides per user)

In [11]:
%%sql
select round(count(distinct order_rk) / count(distinct party_rk)::numeric, 2) as rpu from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('6.21'),)]

- Поездок на пользователя: 6.21

#### Количество поездок с кэшбэком

In [12]:
%%sql
with cashback_rides as (
    select
    count(order_rk) filter (where loyalty_accrual_rub_amt > 0) as rides_with_rub_cashback,
    count(order_rk) filter (where loyalty_accrual_bns_amt > 0) as rides_with_bns_cashback,
    count(order_rk) filter (where loyalty_accrual_rub_amt = 0 and loyalty_accrual_bns_amt = 0) as rides_without_cashback,
    count(order_rk) as total_rides
    from rides
)

select
    total_rides,
    rides_with_rub_cashback, round((rides_with_rub_cashback * 1.0 / total_rides)::numeric, 3) as rides_with_rub_cashback_share,
    rides_with_bns_cashback, round((rides_with_bns_cashback * 1.0 / total_rides)::numeric, 3) as rides_with_bns_cashback_share,
    rides_without_cashback, round((rides_without_cashback * 1.0 / total_rides)::numeric, 3) as rides_without_cashback_share
from cashback_rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(357110, 338585, Decimal('0.948'), 17671, Decimal('0.049'), 856, Decimal('0.002'))]

- Поездки с кэшбэком в рублях: 94,8%
- Поездки с кэшбэком в бонусах (использовалась кредитная карта): 4,9%
- Поездки без кэшбэка: 0,2%

---

### Метрики вовлечённости

#### DAU

In [13]:
%%sql
select
    date(local_book_start_dttm) as date,
    count(distinct party_rk) as dau
from rides
where date(local_book_start_dttm) between '2024-08-01' and '2024-08-10'
group by date
order by date
limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[(datetime.date(2024, 8, 1), 1668),
 (datetime.date(2024, 8, 2), 2016),
 (datetime.date(2024, 8, 3), 1793),
 (datetime.date(2024, 8, 4), 1703),
 (datetime.date(2024, 8, 5), 2020),
 (datetime.date(2024, 8, 6), 2045),
 (datetime.date(2024, 8, 7), 2080),
 (datetime.date(2024, 8, 8), 2004),
 (datetime.date(2024, 8, 9), 2320),
 (datetime.date(2024, 8, 10), 1800)]

#### Average DAU

In [14]:
%%sql
with dau_metrics as (
    select
        date(local_book_start_dttm) as date,
        count(distinct party_rk) as dau
    from rides
    group by date
    order by date
)

select round(avg(dau), 0) as average_dau from dau_metrics;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('1399'),)]

- Среднее количество ежедневных активных пользователей за весь период: 1399

#### WAU

In [15]:
%%sql
select
    date_trunc('week', local_book_start_dttm)::date as week,
    count(distinct party_rk) as wau
from rides
group by week
order by week
limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[(datetime.date(2024, 4, 15), 3),
 (datetime.date(2024, 4, 22), 22),
 (datetime.date(2024, 4, 29), 27),
 (datetime.date(2024, 5, 6), 28),
 (datetime.date(2024, 5, 13), 359),
 (datetime.date(2024, 5, 20), 1042),
 (datetime.date(2024, 5, 27), 1624),
 (datetime.date(2024, 6, 3), 2533),
 (datetime.date(2024, 6, 10), 4391),
 (datetime.date(2024, 6, 17), 4700)]

#### Average WAU

In [16]:
%%sql
with wau_metrics as (
    select
        date_trunc('week', local_book_start_dttm)::date as week,
        count(distinct party_rk) as wau
    from rides
    group by week
    order by week
)

select round(avg(wau), 0) as average_wau from wau_metrics;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('5571'),)]

- Среднее количество еженедельных активных пользователей за весь период: 5571

#### MAU

In [17]:
%%sql
select
    date_trunc('month', local_book_start_dttm)::date AS month_d,
    count(distinct party_rk) as mau
from rides
group by month_d
order by month_d
limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
7 rows affected.


[(datetime.date(2024, 4, 1), 27),
 (datetime.date(2024, 5, 1), 2153),
 (datetime.date(2024, 6, 1), 11737),
 (datetime.date(2024, 7, 1), 17674),
 (datetime.date(2024, 8, 1), 25326),
 (datetime.date(2024, 9, 1), 25034),
 (datetime.date(2024, 10, 1), 13452)]

#### Average MAU

In [18]:
%%sql
with mau_metrics as (
    select
        date_trunc('month', local_book_start_dttm)::date AS month_d,
        count(distinct party_rk) as mau
    from rides
    group by month_d
    order by month_d
)

select round(avg(mau), 0) as average_mau from mau_metrics;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('13629'),)]

- Среднее количество ежемесячных активных пользователей за весь период: 13629

---

### Финансовые метрики

#### Revenue

In [19]:
%%sql
select round(sum(nominal_price_rub_amt)::numeric, 2) as revenue_rub from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('45696200.00'),)]

- Выручка сервиса: 45,7 млн. рублей

#### ARPU

In [20]:
%%sql
select round((sum(nominal_price_rub_amt) / count(distinct party_rk))::numeric, 2) as arpu from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('794.24'),)]

- ARPU: 794,2 рублей

#### AOV (average check)

In [21]:
%%sql
select
    round(avg(nominal_price_rub_amt)::numeric, 2) as avg_price_per_ride_rub,
    round((percentile_cont(0.5) within group (order by nominal_price_rub_amt))::numeric, 2) as median_price_per_ride_rub
from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('128.09'), Decimal('105.92'))]

- Средняя стоимость поездки: 128,09 рублей
- Медианная стоимость поездки: 105,92 рублей

_Это говорит о наличии дорогих поездок (правый хвост распределения)._

#### Average cashback

In [22]:
%%sql
select
    round(avg(loyalty_accrual_rub_amt) filter (where loyalty_accrual_rub_amt > 0) :: numeric, 2) as avg_rub_cashback,
    round(avg(loyalty_accrual_bns_amt) filter (where loyalty_accrual_bns_amt > 0) :: numeric, 2) as avg_bns_cashback
from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(Decimal('33.66'), Decimal('57.52'))]

- Средний кэшбэк для поездок с оплатой по дебетовой карте: 33,66
- Средний кэшбэк для поездок с оплатой по кредитной карте: 57,52

#### Средний чек по типу оплаты (сегментация по типу оплаты)

In [23]:
%%sql
with type_cashback as (
    select
        order_rk,
        case
            when loyalty_accrual_rub_amt > 0 then 'debet'
            when loyalty_accrual_bns_amt > 0 then 'credit'
            else 'unknown'
        end as pay_type
    from rides
)

select t.pay_type, avg(nominal_price_rub_amt) as avg_price_per_ride
from rides r
    inner join type_cashback t using(order_rk)
group by t.pay_type;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
3 rows affected.


[('credit', 129.23927960667575),
 ('debet', 128.09065465779392),
 ('unknown', 105.39254715732325)]

_Тип оплаты не влияет на среднюю стоимость поездки._

#### Cashback share (доля кэшбэка)

In [24]:
%%sql
select
    sum(loyalty_accrual_rub_amt) as sum_cashback_rub,
    round((sum(loyalty_accrual_rub_amt) / sum(nominal_price_rub_amt))::numeric, 2) as cashback_share_rub,
    sum(loyalty_accrual_bns_amt) as sum_cashback_bns,
    round((sum(loyalty_accrual_bns_amt) / sum(nominal_price_rub_amt))::numeric, 2) as cashback_share_bns
from rides;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
1 rows affected.


[(11399410.0, Decimal('0.25'), 1016470.7, Decimal('0.02'))]

- Доля выплаченного кэшбэка в рублях к выручке: 25%
- Доля выплаченного кэшбэка в бонусах к выручке: 2%

---

### Сегментация пользователей

#### По активности

In [25]:
%%sql
with users_rides as (
    select
        party_rk,
        count(order_rk) as rides,
        sum(nominal_price_rub_amt) as revenue
    from rides
    group by party_rk
),
segm as (
    select *,
        case
            when rides = 1 then 'one-time (1)'
            when rides between 2 and 3 then 'casual (2-3)'
            when rides between 4 and 9 then 'regular (4-9)'
            else 'heavy (10+)'
        end as segment
    from users_rides
),
total_stats as (
    select count(distinct party_rk) as total_users, sum(nominal_price_rub_amt) as total_revenue from rides
)

select
    segment,
    count(party_rk) as users,
    round(count(party_rk) / (select total_users from total_stats)::numeric, 3) as user_share,
    sum(rides) as total_rides_per_segment,
    sum(revenue) as revenue,
    round((sum(revenue) / (select total_revenue from total_stats))::numeric, 3) as revenue_share,
    round(avg(rides), 2) as avg_rides_per_segment
from segm
group by segment
order by revenue_share;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
4 rows affected.


[('one-time (1)', 22862, Decimal('0.397'), Decimal('22862'), 4016752.8, Decimal('0.088'), Decimal('1.00')),
 ('casual (2-3)', 14971, Decimal('0.260'), Decimal('35366'), 5289005.0, Decimal('0.116'), Decimal('2.36')),
 ('regular (4-9)', 11080, Decimal('0.193'), Decimal('63865'), 8547966.0, Decimal('0.187'), Decimal('5.76')),
 ('heavy (10+)', 8619, Decimal('0.150'), Decimal('235017'), 27889788.0, Decimal('0.610'), Decimal('27.27'))]

- Heavy пользователи (с 10 и более поездками) дают **61% всей выручки**, хотя их только 15%.
- One-time пользователи (их 39,7%) — дают лишь 8,8% выручки.

#### Топ-10 прибыльных пользователей

In [26]:
%%sql
select
    party_rk,
    count(order_rk) as rides,
    sum(nominal_price_rub_amt) as revenue,
    rank() over (order by sum(nominal_price_rub_amt) desc) as rank
from rides
group by party_rk
limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[(386134592, 343, 54525.797, 1),
 (910211081, 338, 47679.438, 2),
 (759042976, 268, 45547.33, 3),
 (284170812, 190, 36366.44, 4),
 (636557947, 313, 34724.72, 5),
 (231464761, 287, 33637.645, 6),
 (666474893, 191, 32527.781, 7),
 (528704373, 157, 31377.664, 8),
 (391309281, 299, 29861.79, 9),
 (976560193, 167, 29815.598, 10)]

_Активные пользователи и правда прибыльные — стоит рассмотреть программы лояльности или таргет-акции для них._

#### По полу

In [27]:
%%sql
select
    gender_cd,
    count(distinct party_rk) as users,
    count(order_rk) as total_rides,
    round(avg(distance_km)::numeric, 3) as avg_distance_km,
    round(avg(nominal_price_rub_amt)::numeric, 2) as avg_price,
    round(avg(trip_duration_min)::numeric, 2) as avg_trip_duration_min,
    round(avg(cashback_rate_rub)::numeric, 2) as avg_cashback_rub,
    round(avg(cashback_rate_bns)::numeric, 2) as avg_cashback_bns
from rides
group by gender_cd
order by 2 desc;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
3 rows affected.


[('M', 46372, 301828, Decimal('2.429'), Decimal('126.24'), Decimal('11.21'), Decimal('0.23'), Decimal('0.03')),
 ('F', 10719, 52624, Decimal('2.446'), Decimal('139.16'), Decimal('12.76'), Decimal('0.25'), Decimal('0.02')),
 ('U', 441, 2658, Decimal('1.843'), Decimal('119.25'), Decimal('9.90'), Decimal('0.23'), Decimal('0.00'))]

_Средняя длительность поездки и средняя стоимость поездки у женщин выше, чем у мужчин при относительно равной средней дистанции._

#### По количеству поездок на пользователя

In [28]:
%%sql
select gender_cd, round(count(order_rk)::numeric / count(distinct party_rk), 2) as rpu_by_gender
from rides
group by gender_cd;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
3 rows affected.


[('F', Decimal('4.91')), ('M', Decimal('6.51')), ('U', Decimal('6.03'))]

- Количество поездок на пользователя у мужчин: 6,51
- Количество поездок на пользователя у женщин: 4,91

#### По регионам

In [29]:
%%sql
select
    lvn_state_nm as region,
    count(*) as total_rides,
    count(distinct party_rk) as users,
    round(avg(distance_km)::numeric, 3) as avg_distance_km,
    round(avg(nominal_price_rub_amt)::numeric, 2) as avg_price,
    round(avg(trip_duration_min)::numeric, 2) as avg_trip_duration_min,
    sum(nominal_price_rub_amt) as revenue,
    rank() over (order by sum(nominal_price_rub_amt)::numeric desc) as rank
from rides
where lvn_state_nm <> 'Unknown'
group by lvn_state_nm
limit 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
10 rows affected.


[('Г МОСКВА', 82451, 10748, Decimal('2.056'), Decimal('131.21'), Decimal('9.82'), 10820066.0, 1),
 ('МОСКОВСКАЯ ОБЛ', 36173, 5451, Decimal('2.174'), Decimal('131.28'), Decimal('10.51'), 4749524.0, 2),
 ('Г САНКТ-ПЕТЕРБУРГ', 34205, 5027, Decimal('2.636'), Decimal('127.35'), Decimal('10.77'), 4356702.5, 3),
 ('СВЕРДЛОВСКАЯ ОБЛ', 26486, 2997, Decimal('2.294'), Decimal('104.90'), Decimal('11.23'), 2778439.5, 4),
 ('КРАСНОДАРСКИЙ КРАЙ', 14613, 1988, Decimal('2.982'), Decimal('128.20'), Decimal('12.52'), 1873392.5, 5),
 ('РЕСП ТАТАРСТАН', 8743, 1610, Decimal('2.650'), Decimal('115.08'), Decimal('12.80'), 1006165.56, 6),
 ('РЕСП БАШКОРТОСТАН', 7609, 1430, Decimal('2.284'), Decimal('123.39'), Decimal('12.66'), 938868.8, 7),
 ('НИЖЕГОРОДСКАЯ ОБЛ', 6879, 1327, Decimal('2.329'), Decimal('129.00'), Decimal('12.08'), 887357.94, 8),
 ('НОВОСИБИРСКАЯ ОБЛ', 6317, 1055, Decimal('2.111'), Decimal('101.35'), Decimal('11.48'), 640213.25, 9),
 ('РЕСП УДМУРТСКАЯ', 4963, 865, Decimal('3.410'), Decimal('128.61'), Decimal('13.77'), 638285.1, 10)]

#### По моделям самокатов

In [30]:
%%sql
select
    transport_model,
    count(*) as total_rides,
    round(avg(distance_km)::numeric, 3) as avg_distance_km,
    round(avg(nominal_price_rub_amt)::numeric, 2) as avg_price,
    round(avg(trip_duration_min)::numeric, 2) as avg_trip_duration_min,
    round(avg(cashback_rate_rub)::numeric, 2) as avg_cashback_rub
from rides
group by transport_model
order by 2 desc;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
14 rows affected.


[('SL', 218848, Decimal('2.445'), Decimal('130.39'), Decimal('10.84'), Decimal('0.23')),
 ('Max', 62017, Decimal('2.633'), Decimal('131.82'), Decimal('12.83'), Decimal('0.23')),
 ('E', 61195, Decimal('2.182'), Decimal('112.22'), Decimal('11.55'), Decimal('0.23')),
 ('Urban E-Bike', 3909, Decimal('2.589'), Decimal('130.24'), Decimal('10.89'), Decimal('0.27')),
 ('ES400', 3706, Decimal('2.381'), Decimal('138.34'), Decimal('13.98'), Decimal('0.26')),
 ('ES200', 2918, Decimal('2.011'), Decimal('143.32'), Decimal('13.17'), Decimal('0.24')),
 ('ES100', 2425, Decimal('2.110'), Decimal('155.17'), Decimal('15.98'), Decimal('0.28')),
 ('MK', 952, Decimal('2.386'), Decimal('152.76'), Decimal('15.75'), Decimal('0.22')),
 ('ES-08S', 736, Decimal('2.649'), Decimal('192.42'), Decimal('18.41'), Decimal('0.24')),
 ('Techno', 200, Decimal('1.826'), Decimal('143.10'), Decimal('12.55'), Decimal('0.25')),
 ('ES', 91, Decimal('2.332'), Decimal('172.36'), Decimal('18.90'), Decimal('0.30')),
 ('LB', 82, Decimal('2.451'), Decimal('144.17'), Decimal('14.50'), Decimal('0.23')),
 ('B1D', 30, Decimal('2.060'), Decimal('232.16'), Decimal('20.26'), Decimal('0.36')),
 ('Lucky', 1, Decimal('0.295'), Decimal('70.00'), Decimal('23.52'), Decimal('0.10'))]

---

### Когортный анализ по дате первой поездки (месяц первой поездки)

In [31]:
%%sql
with first_ride as (
    select
        party_rk,
        min(date(local_book_start_dttm)) as first_date
    from rides
    group by party_rk
), cohort as (
    select
        party_rk,
        date_trunc('month', first_date)::date as cohort_month
    from first_ride
), activity as (
    select
        r.party_rk,
        date_trunc('month', r.local_book_start_dttm)::date as activity_month,
        c.cohort_month,
        extract(
            month from age(date_trunc('month', r.local_book_start_dttm),
                           c.cohort_month)
        ) as cohort_age
    from rides r
    join cohort c using (party_rk)
), cohort_counts as (
    select cohort_month, count(distinct party_rk) as cohort_size
    from activity
    where cohort_age = 0
    group by cohort_month
)

select
    a.cohort_month,
    a.cohort_age,
    a.users,
    round(a.users::numeric / c.cohort_size, 4) as retention
from (
    select cohort_month, cohort_age, count(distinct party_rk) as users
    from activity
    group by cohort_month, cohort_age
) a
join cohort_counts c using (cohort_month)
order by cohort_month, cohort_age;

 * postgresql+psycopg2://postgres:***@localhost:5432/kicksharing
28 rows affected.


[(datetime.date(2024, 4, 1), Decimal('0'), 27, Decimal('1.0000')),
 (datetime.date(2024, 4, 1), Decimal('1'), 15, Decimal('0.5556')),
 (datetime.date(2024, 4, 1), Decimal('2'), 13, Decimal('0.4815')),
 (datetime.date(2024, 4, 1), Decimal('3'), 17, Decimal('0.6296')),
 (datetime.date(2024, 4, 1), Decimal('4'), 13, Decimal('0.4815')),
 (datetime.date(2024, 4, 1), Decimal('5'), 14, Decimal('0.5185')),
 (datetime.date(2024, 4, 1), Decimal('6'), 7, Decimal('0.2593')),
 (datetime.date(2024, 5, 1), Decimal('0'), 2138, Decimal('1.0000')),
 (datetime.date(2024, 5, 1), Decimal('1'), 992, Decimal('0.4640')),
 (datetime.date(2024, 5, 1), Decimal('2'), 856, Decimal('0.4004')),
 (datetime.date(2024, 5, 1), Decimal('3'), 713, Decimal('0.3335')),
 (datetime.date(2024, 5, 1), Decimal('4'), 612, Decimal('0.2862')),
 (datetime.date(2024, 5, 1), Decimal('5'), 389, Decimal('0.1819')),
 (datetime.date(2024, 6, 1), Decimal('0'), 10732, Decimal('1.0000')),
 (datetime.date(2024, 6, 1), Decimal('1'), 5087, Decimal('0.4740')),
 (datetime.date(2024, 6, 1), Decimal('2'), 3866, Decimal('0.3602')),
 (datetime.date(2024, 6, 1), Decimal('3'), 3126, Decimal('0.2913')),
 (datetime.date(2024, 6, 1), Decimal('4'), 1926, Decimal('0.1795')),
 (datetime.date(2024, 7, 1), Decimal('0'), 11714, Decimal('1.0000')),
 (datetime.date(2024, 7, 1), Decimal('1'), 4492, Decimal('0.3835')),
 (datetime.date(2024, 7, 1), Decimal('2'), 3185, Decimal('0.2719')),
 (datetime.date(2024, 7, 1), Decimal('3'), 1897, Decimal('0.1619')),
 (datetime.date(2024, 8, 1), Decimal('0'), 16242, Decimal('1.0000')),
 (datetime.date(2024, 8, 1), Decimal('1'), 5260, Decimal('0.3239')),
 (datetime.date(2024, 8, 1), Decimal('2'), 2682, Decimal('0.1651')),
 (datetime.date(2024, 9, 1), Decimal('0'), 12837, Decimal('1.0000')),
 (datetime.date(2024, 9, 1), Decimal('1'), 2709, Decimal('0.2110')),
 (datetime.date(2024, 10, 1), Decimal('0'), 3842, Decimal('1.0000'))]

**Анализ:** в ранних когортах (май-июнь) почти половина пользователей возвращалась через месяц после регистрации. Начиная с июля, retention снизился до 38%, а в августовской когорте через месяц вернулось всего 32% пользователей. Тренд указывает на снижение вовлечённости новых пользователей с течением времени — необходимо усиление стимулов для повторной поездки: промокод может увеличить retention.

---

### Выводы

- Данные загружены в реляционную БД.
- Рассчитаны ключевые показатели и метрики сервиса.
- Выручка распределена неравномерно — на **heavy пользователей** (10 и более поездок) приходится больше половины дохода (**61%** всей выручки).
- Мужчины и женщины совершают разное количество поездок за период пользования сервисом (за сезон), что подтверждает выводы из EDA.
- На 99,8% поездок был начислен кэшбэк либо в рублях, либо в бонусах в зависимости от типа оплаты.
- Стоимость поездки не зависит от типа оплаты.
- Постепенное снижение retention со временем указывает на необходимость дополнительных механик удержания пользователей.